File: go/traits.go
package asyncapiclient

const unresolvedTraitField = "x-ob-asyncapi-unresolved-trait"

// applyDocumentTraits implements AsyncAPI 3.0's Traits Merge Mechanism on
// operation and message objects. Traits apply in declaration order using
// JSON Merge Patch, then the target applies last so it always wins.
func applyDocumentTraits(root map[string]any) {
	if operations, ok := objectMap(root["operations"]); ok {
		for name, raw := range operations {
			if operation, ok := objectMap(raw); ok {
				if resolved, ok := resolveTraitTarget(operation, root, nil); ok {
					operations[name] = applyTraits(resolved, root)
				}
			}
		}
	}
	visitMessageMap(root, root["channels"])
	if components, ok := objectMap(root["components"]); ok {
		visitMessages(root, components["messages"])
	}
}

func visitMessageMap(root map[string]any, raw any) {
	channels, ok := objectMap(raw)
	if !ok {
		return
	}
	for name, value := range channels {
		channel, ok := objectMap(value)
		if !ok {
			continue
		}
		if resolved, ok := resolveTraitTarget(channel, root, nil); ok {
			visitMessages(root, resolved["messages"])
			channels[name] = resolved
		}
	}
}

func visitMessages(root map[string]any, raw any) {
	messages, ok := objectMap(raw)
	if !ok {
		return
	}
	for name, value := range messages {
		message, ok := objectMap(value)
		if !ok {
			continue
		}
		if resolved, ok := resolveTraitTarget(message, root, nil); ok {
			messages[name] = applyTraits(resolved, root)
		}
	}
}

func applyTraits(target map[string]any, root map[string]any) map[string]any {
	declared, ok := target["traits"].([]any)
	if !ok {
		return target
	}
	var inherited any = map[string]any{}
	var unresolved string
	for _, raw := range declared {
		trait, ok := objectMap(raw)
		if !ok {
			continue
		}
		resolved, ok := resolveTraitTarget(trait, root, nil)
		if !ok {
			if ref, refOK := trait["$ref"].(string); refOK && unresolved == "" {
				unresolved = ref
			}
			continue
		}
		inherited = mergePatch(inherited, resolved)
	}
	own := cloneMap(target)
	delete(own, "traits")
	merged := mergePatch(inherited, own).(map[string]any)
	if unresolved != "" {
		merged[unresolvedTraitField] = unresolved
	}
	return merged
}

func resolveTraitTarget(value map[string]any, root map[string]any, visited map[string]bool) (map[string]any, bool) {
	ref, hasRef := value["$ref"].(string)
	if !hasRef {
		return cloneMap(value), true
	}
	if visited == nil {
		visited = map[string]bool{}
	}
	if visited[ref] {
		return nil, false
	}
	visited[ref] = true
	target, ok := objectMap(resolveJSONPointer(root, ref))
	if !ok {
		return nil, false
	}
	return resolveTraitTarget(target, root, visited)
}

func mergePatch(target, patch any) any {
	patchObject, ok := objectMap(patch)
	if !ok {
		return cloneValue(patch)
	}
	result := map[string]any{}
	if targetObject, ok := objectMap(target); ok {
		result = cloneMap(targetObject)
	}
	for key, value := range patchObject {
		if value == nil {
			delete(result, key)
			continue
		}
		result[key] = mergePatch(result[key], value)
	}
	return result
}

func objectMap(value any) (map[string]any, bool) {
	result, ok := value.(map[string]any)
	return result, ok
}

func cloneMap(value map[string]any) map[string]any {
	result := make(map[string]any, len(value))
	for key, member := range value {
		result[key] = cloneValue(member)
	}
	return result
}

func cloneValue(value any) any {
	switch typed := value.(type) {
	case map[string]any:
		return cloneMap(typed)
	case []any:
		result := make([]any, len(typed))
		for index, member := range typed {
			result[index] = cloneValue(member)
		}
		return result
	default:
		return value
	}
}

File: typescript/src/traits.ts
import type {
  AsyncAPIChannel,
  AsyncAPIDocument,
  AsyncAPIMessage,
  AsyncAPIOperation,
  AsyncAPIOperationReply,
} from "./asyncapi-types.js";
import { UNRESOLVED_TRAIT_TAG } from "./constants.js";

type JSONObject = Record<string, unknown>;

/**
 * Applies AsyncAPI 3.0's Traits Merge Mechanism after reference resolution.
 * Traits are JSON Merge Patches applied in declaration order, followed by the
 * target itself so a trait never overrides an explicitly declared property.
 */
export function applyDocumentTraits(document: AsyncAPIDocument): AsyncAPIDocument {
  if (document.operations) {
    for (const [key, operation] of Object.entries(document.operations)) {
      document.operations[key] = applyTraits(operation as unknown as JSONObject) as unknown as AsyncAPIOperation;
    }
  }

  visitMessages(document.components?.messages);
  for (const channel of Object.values(document.channels ?? {})) visitChannelMessages(channel);
  for (const operation of Object.values(document.operations ?? {})) {
    visitMessageList(operation.messages);
    visitChannelMessages(operation.channel);
    visitReplyMessages(operation.reply);
  }
  return document;
}

function visitReplyMessages(reply: AsyncAPIOperationReply | undefined): void {
  if (!reply) return;
  visitMessageList(reply.messages);
  visitChannelMessages(reply.channel);
}

function visitChannelMessages(channel: AsyncAPIChannel | undefined): void {
  if (channel) visitMessages(channel.messages);
}

function visitMessages(messages: Record<string, AsyncAPIMessage> | undefined): void {
  if (!messages) return;
  for (const [key, message] of Object.entries(messages)) {
    messages[key] = applyTraits(message as unknown as JSONObject) as unknown as AsyncAPIMessage;
  }
}

function visitMessageList(messages: AsyncAPIMessage[] | undefined): void {
  if (!messages) return;
  for (let index = 0; index < messages.length; index += 1) {
    messages[index] = applyTraits(messages[index] as unknown as JSONObject) as unknown as AsyncAPIMessage;
  }
}

function applyTraits(target: JSONObject): JSONObject {
  const declared = target["traits"];
  if (!Array.isArray(declared)) return target;

  let inherited: unknown = {};
  let unresolved: string | undefined;
  for (const trait of declared) {
    if (!isJSONObject(trait)) continue;
    if (typeof trait["$ref"] === "string") {
      unresolved ??= trait["$ref"];
      continue;
    }
    inherited = mergePatch(inherited, trait);
  }

  const own = { ...target };
  delete own["traits"];
  const merged = mergePatch(inherited, own) as JSONObject;
  if (unresolved) merged[UNRESOLVED_TRAIT_TAG] = unresolved;
  return merged;
}

/** RFC 7396 JSON Merge Patch over JSON-compatible AsyncAPI objects. */
function mergePatch(
  target: unknown,
  patch: unknown,
  mergedPatches = new WeakMap<object, unknown>(),
): unknown {
  if (!isJSONObject(patch)) return cloneGraph(patch);

  // Reference resolution turns recursive schemas into cyclic object graphs.
  // JSON Merge Patch is defined over JSON trees, but applying the same rules to
  // that resolved graph must preserve a back-edge instead of recursively walking
  // it forever.
  const previous = mergedPatches.get(patch);
  if (previous !== undefined) return previous;

  const result: JSONObject = isJSONObject(target) ? cloneGraph(target) as JSONObject : {};
  mergedPatches.set(patch, result);
  for (const [key, value] of Object.entries(patch)) {
    if (value === null) delete result[key];
    else result[key] = mergePatch(result[key], value, mergedPatches);
  }
  return result;
}

function cloneGraph<T>(value: T, clones = new WeakMap<object, unknown>()): T {
  if (value === null || typeof value !== "object") return value;

  const previous = clones.get(value);
  if (previous !== undefined) return previous as T;

  if (Array.isArray(value)) {
    const result: unknown[] = [];
    clones.set(value, result);
    for (const member of value) result.push(cloneGraph(member, clones));
    return result as T;
  }
  if (isJSONObject(value)) {
    const result: JSONObject = {};
    clones.set(value, result);
    for (const [key, member] of Object.entries(value)) result[key] = cloneGraph(member, clones);
    return result as T;
  }
  return value;
}

function isJSONObject(value: unknown): value is JSONObject {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}
